File: src/main/java/june_2025/MyPow.java
package june_2025;

public class MyPow {

    public static void main(String[] args) {
        MyPow myPow = new MyPow();
        double result = myPow.myPow(2.0, 10);
        System.out.println(result);
    }

    public double myPow(double x, int n) {
        if (n == 0) {
            return 1;
        }
        return n < 1 ? 1 / helper(x, Math.abs((long) n)) : helper(x, Math.abs((long) n));
    }

    private double helper(double x, long n) {
        if (n == 1) {
            return x;
        }
        double half = helper(x, n / 2);
        return n % 2 == 0 ? half * half : x * half * half;
    }

}

File: python/second/my_pow.py
class Solution(object):
    # O(log(n)) time | O(log(n)) space
    def myPow(self, x, n):
        """
        :type x: float
        :type n: int
        :rtype: float
        """
        if n == 0:
            return 1
        if x == 1.0:
            return x
        
        return self.helper(x, n) if n > 0 else 1 / self.helper(x, -1 * n)

    def helper(self, x, n):
        if n == 0:
            return 1
        half = self.helper(x, n // 2)
        return half * half if n % 2 == 0 else x * half * half
    

solution = Solution()
# x = 2.00000
# n = 10
x = 2.00000
n = -2
print(solution.myPow(x, n))
